pkg/pki/ca/ca.go

```diff
@@ -56,24 +74,57 @@ type IstioCA struct {
 }
 
 // NewSelfSignedIstioCA returns a new IstioCA instance using self-signed certificate.
-func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioCA, error) {
-	now := time.Now()
-	options := CertOptions{
-		NotBefore:    now,
-		NotAfter:     now.Add(caCertTTL),
-		Org:          org,
-		IsCA:         true,
-		IsSelfSigned: true,
-		RSAKeySize:   caKeySize,
-	}
-	pemCert, pemKey := GenCert(options)
+func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string, namespace string,
+	core corev1.CoreV1Interface) (*IstioCA, error) {
 
+	caSecret, err := core.Secrets(namespace).Get(CASecret, metav1.GetOptions{})
 	opts := &IstioCAOptions{
-		CertTTL:          certTTL,
-		SigningCertBytes: pemCert,
-		SigningKeyBytes:  pemKey,
-		RootCertBytes:    pemCert,
+		CertTTL:   certTTL,
+		Namespace: namespace,
+	}
+	if err != nil {
+		glog.Warningf("Failed to get secret (error: %s), will create one", err)
+
+		now := time.Now()
+		options := CertOptions{
+			NotBefore:    now,
+			NotAfter:     now.Add(caCertTTL),
+			Org:          org,
+			IsCA:         true,
+			IsSelfSigned: true,
+			RSAKeySize:   caKeySize,
+		}
+		pemCert, pemKey := GenCert(options)
+
+		opts.SigningCertBytes = pemCert
+		opts.SigningKeyBytes = pemKey
+		opts.RootCertBytes = pemCert
+
+		// Rewrite the key/cert back to secret so they will be persistent when CA restarts.
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				CACertChainID:  pemCert,
+				CAPrivateKeyID: pemKey,
+				RootCertID:     pemCert,
```

Comment:
Nit: you don't need to save root cert into the secret as it's always identical to the ca cert.